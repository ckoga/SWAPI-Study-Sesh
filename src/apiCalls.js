export const getFilms = () => {
  return fetch('https://swapi.co/api/films/')
    .then(response => {
      if (!response.ok) {
        throw Error('Error fetching films');
      } 
      return response.json()
    });
};