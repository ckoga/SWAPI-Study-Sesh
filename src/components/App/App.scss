@use 'variables'

// if i remove this body selector the background goes white
// even though there are no css rules applied to this selector
body {
}

.App {
  background-color: #333333;
  height: 200vh;
}

.App-header {
  background-color: #4F4F4F;
  color: #F3CA3E;
  border: 2px solid #F3CA3E;
  border-left: none;
  border-right: none;
  // margin-top: 15px;
  width: 100%;
  // position: fixed;
  height: 10%;
}

h1 {
  margin: 10px;
}
