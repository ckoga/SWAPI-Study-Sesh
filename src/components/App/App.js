import React, {Component} from 'react';
import './App.scss';
import '../../_variables.scss';
import WelcomeForm from '../WelcomeForm/WelcomeForm';
import DisplayContainer from '../DisplayContainer/DisplayContainer';
import Header from '../Header/Header';
import {Route} from 'react-router-dom';
import { getFilms } from '../../apiCalls';

class App extends Component {
  constructor() {
    super();
    this.state = {
      movies: [],
      characters: [],
      scroll: '',
      error: '',
      isLoading: true,
      user: {
        name: '',
        quote: '',
        rank: '',
      }
    }
  }

  updateUser = (user) => {
    this.setState({ user })
  }

  componentDidMount() {
    getFilms()
      .then(data => this.setState({ movies: data.results, isLoading: false }))
      .catch(err => {
        console.log(err);
        this.setState({ error: err.message})
      })
  }

  setScrollAndLoading = (movie) => {
    this.setState({ scroll: movie.opening_crawl, isLoading: true, characters: [] })
  }

  fetchHandler = (episode) => {
    const selectedMov = this.state.movies.find(movie => movie.episode_id === episode)
    const tenCharacters = selectedMov.characters.slice(0, 10);

    this.setScrollAndLoading(selectedMov);

    const fetchCharData = () => {
      const fetchedCharacters = tenCharacters.map(character => {
        return fetch(character)
        .then(response => {
          if (!response.ok) {
            throw Error('Error fetching films');
          }
          return response.json()
        })
          .then(charData => getNestedData(charData))
      });
      Promise.all(fetchedCharacters)
        .then(data => this.setState({ characters: data, isLoading: false }))
        .catch(err => {
          console.log(err);
          this.setState({ error: err.message})
        })
    }

    const getNestedData = (charData) => {
      const homeWorld = fetchHomeWorld(charData)
      const species = fetchSpecies(charData)
      const films = fetchFilms(charData)
      return Promise.all([homeWorld, species, ...films])
        .then(data => {
          return createCharacterObj(data, charData.name)})
    }

    const createCharacterObj = (resolvedData, name) => {
      return {
        name: name,
        homeworld: resolvedData[0].name,
        population: resolvedData[0].population,
        species: resolvedData[1].name,
        films: resolvedData.slice(2).map(film => film.title),
      }
    }

    const fetchHomeWorld = (charData) => {
      return fetch(charData.homeworld)
      .then(response => {
        if (!response.ok) {
          throw Error('Error fetching films');
        }
        return response.json()
      })
    }

    const fetchSpecies = (charData) => {
      return fetch(charData.species)
      .then(response => {
        if (!response.ok) {
          throw Error('Error fetching films');
        }
        return response.json()
      })
    }

    const fetchFilms = (charData) => {
      const filmPromises = charData.films.map(film => {
        return fetch(film)
        .then(response => {
          if (!response.ok) {
            throw Error('Error fetching films');
          }
          return response.json()
        })
      });
      return filmPromises;
    }

    fetchCharData();
  }

  render() {
    return (
      <div className="app">
        <Header
          name={this.state.user.name}
          quote={this.state.user.quote}
          rank={this.state.user.rank}
          updateUser={this.updateUser}
        />
        {this.state.error && <h2 className='fetch-error'>{this.state.error}</h2>}
        <Route exact path='/' render={ () =>
          <WelcomeForm
            updateUser={this.updateUser}
          />
        } />
        <Route exact path='/movies' render={ () => {
          return (<DisplayContainer
            data={this.state.movies}
            fetchHandler={this.fetchHandler}
            isMovies={true}
            isLoading={this.state.isLoading}
          />)}
        } />
        <Route path='/movies/:movie_id' render={ ({ match }) => {
          return(<DisplayContainer
            scroll={this.state.scroll}
            data={this.state.characters}
            isLoading={this.state.isLoading}
            />)}
        } />
      </div>
    );
  }
}

export default App;
