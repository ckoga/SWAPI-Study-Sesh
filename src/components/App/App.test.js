import React from 'react';
import App from './App';
import {shallow} from 'enzyme';

import { getFilms } from '../../apiCalls';
jest.mock('../../apiCalls.js');

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    getFilms.mockImplementation(() => {
      return Promise.resolve([{
      "title": "The Empire Strikes Back",
      "episode_id": 5,
      "opening_crawl": "Impressive opening starwars crawl",
      "director": "Irvin Kershner",
      "producer": "Gary Kurtz, Rick McCallum",
      "release_date": "1980-05-17",
      "characters": ["https://swapi.co/api/people/1/"],
      "planets": ["https://swapi.co/api/planets/4/"],
      "starships": ["https://swapi.co/api/starships/15/"],
      "vehicles": ["https://swapi.co/api/vehicles/8/"],
      "species": ["https://swapi.co/api/species/6/"],
      "created": "2014-12-12T11:26:24.656000Z",
      "edited": "2017-04-19T10:57:29.544256Z",
      "url": "https://swapi.co/api/films/2/"
      }])
    })
    wrapper = shallow(<App />);
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should call getFilms after component is mounted', () => {
    expect(getFilms).toHaveBeenCalled();
  })

  it('should update state when updateUser is invoked', () => {
    const mockUser = {
      name: 'Robbie',
      quote: '"Luke. I am your father."',
      rank: 'Ranking: Padawan', };
    const expected = {
      name: 'Robbie',
      quote: '"Luke. I am your father."',
      rank: 'Ranking: Padawan', };
    wrapper.instance().updateUser(mockUser);

    expect(wrapper.state('user')).toEqual(expected);
  })

  it('should update state when setScrollAndLoading is invoked', () => {
    const mockMovie = {
    "title": "The Empire Strikes Back",
    "episode_id": 5,
    "opening_crawl": "Impressive opening starwars crawl",
    "director": "Irvin Kershner",
    "producer": "Gary Kurtz, Rick McCallum",
    "release_date": "1980-05-17",
    "characters": ["https://swapi.co/api/people/1/"],
    "planets": ["https://swapi.co/api/planets/4/"],
    "starships": ["https://swapi.co/api/starships/15/"],
    "vehicles": ["https://swapi.co/api/vehicles/8/"],
    "species": ["https://swapi.co/api/species/6/"],
    "created": "2014-12-12T11:26:24.656000Z",
    "edited": "2017-04-19T10:57:29.544256Z",
    "url": "https://swapi.co/api/films/2/"
    };
    const expected = "Impressive opening starwars crawl";

    wrapper.instance().setScrollAndLoading(mockMovie);

    expect(wrapper.state('scroll')).toEqual(expected);
  })

})
