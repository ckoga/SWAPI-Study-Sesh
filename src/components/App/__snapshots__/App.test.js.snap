// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div
  className="app"
>
  <Header
    name=""
    quote=""
    rank=""
    updateUser={[Function]}
  />
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/movies"
    render={[Function]}
  />
  <Route
    path="/movies/:movie_id"
    render={[Function]}
  />
</div>
`;
